T-8952: Reload Vector when enrichment table changes

paweljw opened the pull request.

kessl commented on engine/vector_enrichment_table.rb:
```diff
@@ -0,0 +1,20 @@
+# Checks whether the enrichment table has changed
+# Returns a hash of the file if it exists, otherwise nil
+# Used by enrichment_table_watcher.rb to reload vector if the enrichment table has changed
+class VectorEnrichmentTable
```
```suggestion
class VectorEnrichmentTableUpdater
```

paweljw replied:
I changed it quite a bit. With `VectorEnrichmentTable.new(...).validate` and `VectorEnrichmentTable.different?` I think the current name works better :)

paweljw commented:
@curusarn ah yeah, fair shout. I’ll extract the reload logic, make it more robust (e.g. checking if the file isn’t 0-length or something), and merge the enrichment table update with config update. 

Idea being, the ping+update common supervisord service will tick every 15s, doing ping every 30s, and doing the enrichment table check every 15s. Does that track?

paweljw commented:
> Could you make sure we'll never use an enrichment table that would crash vector? E.g. run "vector validate" on some minimal vector config + the enrichment table.
I’ll check what happens with a broken csv passed in, but vector validate afair didn’t load in the enrichment table csv 
> Could you make sure we validate the entire config together with enrichment table before reloading vector?
I think that again will be making sure the csv itself is valid, because the table strictly speaking isn’t in the config. It’s descriptor is, but that’s static and already part of the validation. 

That’s a reasonable direction though, I’ll do a little research into how to validate the enrichment table correctness itself. 🫡

paweljw commented:
@curusarn I dropped the separate script, added more validation to enrichment table, worked around a bug that'd be caused by no dockerprobe, and squished it all into updater. Let me know what you think.

curusarn reviewed (changes requested):
Thanks @paweljw!

I was expecting the enrichment table validation would plug into the existing validation here: 
https://github.com/BetterStackHQ/collector/pull/18/files#diff-a683eb9095bba9b10a2bf60548f06bf923fd29314373e97ac122f261e30c0c02R164

So that we would always run `vector validate` on the full complete config to be used by vector before moving the configs to path vector uses and reloading vector.

If I read the flow of the code correctly we only have one path for enrichment table: "/enrichment/docker-mappings.csv".
If we generate an invalid enrichment table it will be immediately saved to that path. Once we generate an invalid enrichment table, it looks like vector validate will never pass so we'll never use any new upstream vector configs.

If we already generated an invalid enrichment table and vector restarts for any reason it looks like vector will try to load the enrichment table and fail.



In contrast, when we generate invalid kubernetes discovery configs we make sure we'll keep using a fallback version or a previous version instead. This ensures we will still use upstream config updates and we don't stop sending data. (https://github.com/BetterStackHQ/collector/blob/main/engine/kubernetes_discovery.rb#L245, https://github.com/BetterStackHQ/collector/blob/main/engine/utils.rb#L94)

We always keep the `/vector-config/current` valid and only replace it with new fully validated config. I would handle the enrichment table the same way.
This ensures that randomly restarting vector doesn't cause it to crash.

I'm assuming invalid enrichment tables have potential to crash vector. Please correct me if I'm wrong. 🙏 

Anything I'm missing? 🙏

curusarn commented:
<!-- linear:replyTo:3084669601 -->

> If you start a container which creates a new volume, and the container has files or directories in the directory to be mounted such as `/app/`, Docker copies the directory's contents into the volume.

curusarn commented:
<!-- linear:replyTo:3084669601 -->

But I'm not strongly opposed to populating from updater either. I didn't realize it's different than with our other configs because of the shared volume.

paweljw commented:
@curusarn changes implemented as we discussed.

curusarn reviewed (approved):
Thank you @paweljw!

jurajmasar commented on docker-compose.yml:
```diff
@@ -3,7 +3,7 @@ services:
       context: .
       dockerfile: Dockerfile
-    image: betterstack/collector:latest
+    # image: betterstack/collector:latest
```
@paweljw was this intentional, please?

paweljw replied:
😱 no, https://github.com/BetterStackHQ/collector/pull/19

jurajmasar commented on docker-compose.yml:
```diff
@@ -32,7 +32,7 @@ services:
       context: .
       dockerfile: Dockerfile.beyla
-    image: betterstack/collector-beyla:latest
+    # image: betterstack/collector-beyla:latest
```
@paweljw was this intentional, please?
